Add examples

wusto opened the pull request:
Examples for the following use-cases are added:

- symmetric encryption (incl. authenticated encryption)
- CA creation, CSR signing and CSR creation
- ECIES encryption and decryption
- HASH value calculation
- Key derivation
- Key creation (ECC, Ed-ECC, RSA)
- MAC calculation and verification
- RSA encryption and decryption (incl. padding)
- Signing and verifying of messages using RSA, ECDSA, EdDSA

Commit:
DigestTypes: Move from openssl to mococrw namespace

The DigestTypes enum is a OpenSSL independent enum, which was defined in
the openssl namespace. As it is independent it is moved to the mococrw
namespace.

For API compatibilty reasons a
using DigestTypes = mococrw::DigestTypes;

statement is added to the openssl namespace.

Commit:
Fix typos

Commit:
CRL-tests: Remove duplicate code

Commit:
Fix: add error detection to fromHex

The function fromHex didn't recognize failures during
string->std::vector<uint8_t> parsing.
It throws MoCOCrWException from now on, if the hex string can't be
parsed.

Commit:
Update docu of AESCipherBuilder::getAuthTag

Commit:
Add examples

Examples for the following use-cases are added:
- symmetric encryption (incl. authenticated encryption)
- CA creation, CSR signing and CSR creation
- ECIES encryption and decryption
- HASH value calculation
- Key derivation
- Key creation (ECC, Ed-ECC, RSA)
- MAC calculation and verification
- RSA encryption and decryption (incl. padding)
- Signing and verifying of messages using RSA, ECDSA, EdDSA

Commit:
hash: Add new instantiation function

Add a new function to retrieve a Hash object based on a given
DigestTypes.

neverpanic commented on src/kdf.cpp:
```diff
@@ -21,10 +21,9 @@
-KeyDerivationFunction::~KeyDerivationFunction() = default;
-
-
+using namespace openssl;
```
Please move the changelog entry for this change into this commit. I don't want to risk loosing the changelog entry if we have to revert the rest of the changes but keep this commit.

neverpanic commented on src/mococrw/private/IOUtils.h:
```diff
@@ -65,13 +65,28 @@ mococrw::X509Certificate loadCertFromDERFile(const std::string &filename)
+    return openSSLObjectFromFile(mococrw::AsymmetricPublicKey::readPublicKeyFromPEM, filename);
 }
 
+mococrw::AsymmetricPrivateKey loadPrivkeyFromFile(const std::string &filename, const std::string &password)
```
Again, please adjust the commit message – this change does not only make IOUtils.h a private header, it also extends IOUtils.h with added functionality, but the commit message makes this sound like there isn't a content change in there at all.

neverpanic commented on src/mococrw/private/IOUtils.h:
```diff
@@ -65,13 +65,28 @@ mococrw::X509Certificate loadCertFromDERFile(const std::string &filename)
+{
+    std::ofstream file{filename, std::ofstream::out};
+    if (!file.good()) {
+	std::string errorMsg{"Cannot open file for writing PEM."};
```
Indentation is inconsistent (possibly mixing tabs and spaces?). Also, in other error cases in this file, we're throwing a `std::runtime_error`, so I'd prefer if we did the same thing here.

neverpanic commented on src/util.cpp:
```diff
@@ -41,9 +41,27 @@ std::string toHex(const std::vector<uint8_t> &data) {
     binary.reserve(hexData.size()/2);
-    for (size_t i=0; i<hexData.length(); i+=2) {
+    size_t startPos = 0;
+    if (hexData.rfind("0x") == 0) {
```
src/mococorw/util.h is a public header, so this is a change of behavior in a public API. This should at least come with a changelog entry. We should, however, really also add a documentation comment to the header explaining this case and that the function can throw.

Additionally, I think the special handling for the `0x` prefix is unnecessary, from `strtoul(3)`:
> The string may begin with an arbitrary amount of white space (as determined by isspace(3)) followed by a single optional '+' or '-' sign. If base is  zero or 16, the string may then include a "0x" prefix, and the number will be read in base 16

wusto replied:
That's not true for our case as we always take the substrings from the "complete" number. In consequence we extract "0x" as the first two chars to convert and this is an invalid string for strtoul.

neverpanic replied:
Ah, good point – we should probably also support `0X` then (even if it's uncommon). The point about the header documentation and changelog entry is still valid, though.

neverpanic commented on src/util.cpp:
```diff
@@ -41,9 +41,27 @@ std::string toHex(const std::vector<uint8_t> &data) {
+            throw MoCOCrWException(error_string);
+        }
+
+        if (endptr == encodedByte.c_str()) {
```
I think this check should be rewritten to `if (*endptr == '\0')`, in reference to the following section of the manpage (highlight mine):

> If  endptr is not NULL, strtoul() stores the address of the first invalid character in *endptr.  If there were no digits at all, strtoul() stores the original value of nptr in *endptr (and returns 0). **In particular, if \*nptr is not '\0' but \*\*endptr is '\0' on return, the entire string is valid.**

This would catch the case of a string that doesn't have an even length, which I don't think the current code handles.

wusto replied:
I did some testing. 
Odd strings are always valid. So 0x1 is as valid as 0x01.
The reasons are:
- string.substr returns "The substring is the portion of the object that starts at character position pos and spans len characters (or until the end of the string, whichever comes first)."
- strtoul "converts the initial part of the string in nptr to an unsigned long int" <- there is no length limitation. I guess that it just needs to fit into a unsigned long int

I adapted the checks according to the documentation.

neverpanic commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+static const SymmetricCipherMode plainOperationMode = SymmetricCipherMode::CBC;
+static const SymmetricCipherPadding padding = SymmetricCipherPadding::PKCS;
+static const std::vector<uint8_t> aesKey = utility::fromHex("000102030405060708090a0b0c0d0e0f");
+static const std::vector<uint8_t> authData = utility::fromHex("beefdead");
```
```suggestion
static const std::vector<uint8_t> associatedData = utility::fromHex("beefdead");
```

Would rename so that it's clear from the "**A**uthenticated **E**ncryption with **A**ssociated **D**ata" acronym that this is the latter part.

garw replied:
yep. When I first read it, I had the impression that this should be the authTag.

neverpanic commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+
+    auto authenticatedDecryptor = decryptorBuilder.buildAuthenticatedEncryptor();
+    authenticatedDecryptor->addAssociatedData(authData);
+    authenticatedDecryptor->setAuthTag(aesEncryptResult.authTag);
```
I'd re-order the `->setAuthTag()` call just before `->finish()`, since that's where it is usually called in streaming decryption.

neverpanic commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+        authenticatedDecryptor->update(aesEncryptResult.ciphertext);
+        authenticatedDecryptor->finish();
+    }  catch (MoCOCrWException &e) {
+        std::cout << "Decryption failed!" << std::endl;
```
```suggestion
        std::cout << "Decryption or auth tag verification failed!" << std::endl;
```

I'd extend the message with auth tag failure, just so it's immediately clear how an authentication failure would show up in our API.

neverpanic commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+int main(void) {
+    /* Authenticated encryption and decryption */
+    auto aeResult = aesAuthenticatedEncryption();
+    aesAuthenticatedDecryption(aeResult);
```
Might be nice to also add the `plaintext != decryptionResult` check here?

wusto replied:
This is implicitly checked by verifying the auth tag so I didn't add it here.
Do you think that it will help the reader of this example?

neverpanic replied:
A right, that makes sense – in that case I guess we can skip it.

garw replied:
It is but it isn't. The crypto verified this but that doesn't mean that our code does the right thing…

neverpanic commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,213 @@
+                                                                         .build());
+static const BasicConstraintsExtension rootCertConstraint{true, 1};
+static const BasicConstraintsExtension rootCertSignConstraints{true, 0};
+static const auto rootCaKeyUsage = std::make_unique<KeyUsageExtension>(KeyUsageExtension::Builder{}
```
Does this need to be a `unique_ptr`? It's de-referenced and copied anyway, so we might as well create it using a copy here, i.e.

```c++
static const KeyUsageExtension rootCaKeyUsage{KeyUsageExtension::Builder{}/* ... */.build()};
```

or is there a compilation problem with this solution? The copy constructor doesn't seem to be deleted, so I'd expect this to work fine. The same question applies in a couple of places below, e.g. for `intermediateCaKeyUsage`, but also for `intermediateCaSigningConstraints`, which is a `unique_ptr`, while `rootCertConstraint` isn't.

neverpanic commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,213 @@
+    try {
+        cert = std::make_shared<X509Certificate>(ca->signCSR(*csr));
+    }  catch (openssl::OpenSSLException &e) {
+        std::cerr << "Error reading CSR in PEM format. Please validate  your CSR." << std::endl;
```
```suggestion
        std::cerr << "Error reading CSR in PEM format. Please validate your CSR." << std::endl;
```

neverpanic commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,213 @@
+    try {
+        cert = std::make_shared<X509Certificate>(ca->signCSR(*csr));
+    }  catch (openssl::OpenSSLException &e) {
+        std::cerr << "Error reading CSR in PEM format. Please validate  your CSR." << std::endl;
```
```suggestion
        std::cerr << "Error reading CSR in PEM format. Please validate your CSR." << std::endl;
```

neverpanic commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,148 @@
+    std::vector<uint8_t> mac;
+};
+
+static const std::string pubKeyPem =R"(
```
```suggestion
static const std::string pubKeyPem = R"(
```

neverpanic commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,148 @@
+-----END PUBLIC KEY-----
+)";
+
+static const std::string privKeyPem =R"(
```
```suggestion
static const std::string privKeyPem = R"(
```

neverpanic commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,148 @@
+     * - private key
+     */
+
+    /* Use the default values, which should match here. */
```
Are you actually using the default values? `buildDecryptionContext` seems very similar to `buildEncryptionContext` to me.

neverpanic commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,148 @@
+int main(void)
+{
+    auto eciesData = encrypt_ecies();
+    decrypt_ecies(eciesData);
```
In the AES example, you compared plaintext to decrypted ciphertext. Do you also want to do this here?

wusto replied:
It's the same as for the AEAD encryption. If the MAC verification passes, the decrypted ciphertext should be identical to the plaintext. This is checked in the unit tests.
But if this help the user of this example, I can add it. Please leave a short comment :)

neverpanic replied:
Same as above, fine with me to not check this explicitly.

garw replied:
Same as above: The crypto should do this. Does not mean that our code actually does so.

garw replied:
I guess this is still open?

neverpanic commented on examples/key-example.cpp:
```diff
@@ -0,0 +1,49 @@
+
+void createRsaKey(uint keySize, std::string password)
+{
+    std::shared_ptr<AsymmetricKey::Spec> spec;
+    spec = std::make_shared<RSASpec>(keySize);
```
```suggestion
    std::shared_ptr<AsymmetricKey::Spec> spec = std::make_shared<RSASpec>(keySize);
```

Would just keep this in a single line.

garw replied:
Why even have a ptr?

neverpanic commented on examples/mac-example.cpp:
```diff
@@ -0,0 +1,67 @@
+std::unique_ptr<mococrw::HMAC> doHmac(const DigestTypes &digestType, const std::vector<uint8_t> &key,
+                                      const std::vector<uint8_t>& message)
+{
+    auto hmac =  std::make_unique<mococrw::HMAC>(digestType, key);
```
```suggestion
    auto hmac = std::make_unique<mococrw::HMAC>(digestType, key);
```

neverpanic commented on examples/mac-example.cpp:
```diff
@@ -0,0 +1,67 @@
+
+using namespace mococrw;
+
+std::unique_ptr<mococrw::HMAC> doHmac(const DigestTypes &digestType, const std::vector<uint8_t> &key,
```
```suggestion
std::unique_ptr<mococrw::HMAC> doHmac(DigestTypes digestType, const std::vector<uint8_t> &key,
```

We don't usually pass simple types by reference, and IIRC `DigestTypes` is just an enum. The same applies to verify() and possibly others.

neverpanic commented on examples/mac-example.cpp:
```diff
@@ -0,0 +1,67 @@
+using namespace mococrw;
+
+std::unique_ptr<mococrw::HMAC> doHmac(const DigestTypes &digestType, const std::vector<uint8_t> &key,
+                                      const std::vector<uint8_t>& message)
```
```suggestion
                                      const std::vector<uint8_t> &message)
```
Nitpick: Inconsistency in the position of the ampersand.

neverpanic commented on examples/rsa-example.cpp:
```diff
@@ -0,0 +1,83 @@
+{
+    auto keyPair = std::make_shared<AsymmetricKeypair>(AsymmetricKeypair::generateRSA());
+    std::vector<uint8_t> data = utility::fromHex("deadbeef");
+    DigestTypes oaepDigestType = DigestTypes::SHA512;
+    DigestTypes mgfDigestType = DigestTypes::SHA3_256;
+    std::string oaepLabel = "this is my awesome label";
```
I like that you're showing that we can adjust all these things, but since your example is a bit uncommon (usually the same hash is used for OAEP and the MGF, the label is virtually never modified and always the default), I'd appreciate if you could add a short comment that this a very constructed example and you usually don't have to set these.

neverpanic commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,198 @@
+using namespace mococrw;
+
+std::shared_ptr<MessageSignatureCtx> getRsaSignCtx(std::shared_ptr<AsymmetricPrivateKey> privKey,
+                                                   const DigestTypes &digestType,
```
`const DigestTypes&` vs. `DigestType`, see above.

neverpanic commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,198 @@
+
+std::shared_ptr<MessageSignatureCtx>getEcdsaSignCtx(std::shared_ptr<AsymmetricPrivateKey> privKey,
+                                                    const DigestTypes &digestType,
+                                                    const ECDSASignatureFormat &sigFormat)
```
I think sigFormat is also just an enum, right?

neverpanic commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,198 @@
+
+    /* we can use here the private key, as it also contains the public key.
+     * In MoCOCrW the AsymmetricPrivateKey is a specialisation of AsymmetricPublicKey. Thus
+     * we do an implicit downcast here.
```
Isn't this an upcast, since we're casting to the superclass (i.e. the one above in the type hierarchy)?

neverpanic commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,198 @@
+
+    /* we can use here the private key, as it also contains the public key.
+     * In MoCOCrW the AsymmetricPrivateKey is a specialisation of AsymmetricPublicKey. Thus
+     * we do an implicit downcast here.
```
Same would apply here.

neverpanic commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,198 @@
+
+    /* we can use here the private key, as it also contains the public key.
+     * In MoCOCrW the AsymmetricPrivateKey is a specialisation of AsymmetricPublicKey. Thus
+     * we do an implicit downcast here.
```
And here.

neverpanic commented on CHANGELOG.md:
```diff
@@ -8,16 +8,30 @@ All notable changes to this project will be documented in this file.
 
 * Replace `explicit_bzero(3)` with `OPENSSL_cleanse(3)` to fix compilation on non-glibc systems. (see #75)
 * Replaced manual calculation of hash digest length by library function in ECIES
+* Move DigestTypes from openssl namespace to mococrw namespace
```
This should mention that this is an ABI change.

neverpanic commented on CHANGELOG.md:
```diff
@@ -8,16 +8,30 @@ All notable changes to this project will be documented in this file.
 
 * Exception message in AESCipher (thrown if key has unexpected length)
 * Install target of documentation
+* Add error detection to utility::fromHex
```
Please add a note indicating that `fromHex` can now throw.

neverpanic reviewed (changes requested):
Please check your `catch (ExceptionType& e)` blocks and turn them into `catch (const ExceptionType& e)` blocks.

blenk92 commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+            // Padding mode is optional. Default is SymmetricCipherPadding::PKCS
+            .setPadding(padding)
+            // The IV is optional. Default is a random value
+            .setIV(iv);
```
isn't that something we expect people not to set manually in most cases? if so, I also would just leave this as a comment here.

blenk92 commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+            // Padding mode is optional. Default is SymmetricCipherPadding::PKCS
+            .setPadding(padding)
+            // The IV is optional. Default is a random value
+            .setIV(iv);
```
same here

garw replied:
Maybe we want to demonstrate the random IV workflow here?

blenk92 commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,213 @@
+                                                                 std::unique_ptr<AsymmetricKeypair> &rootCaKey,
+                                                                 std::shared_ptr<X509Certificate> rootCert)
+{
+    auto csr = std::make_unique<CertificateSigningRequest>(CertificateSigningRequest(
```
this unique_ptr is passed as a reference to `rootCaSign()` and then dereferenced in that function. Seems like a "normal" object would also do... Or is there anything I'm missing here?
Same is true for at least some of the other unique_ptrs. If they are not required i like to see them removed to make the code a bit simpler (so people can easily understand whats going on)

blenk92 commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,148 @@
+
+    std::vector<uint8_t> encryptedData;
+    try {
+         encCtx->update(message);
```
```suggestion
        encCtx->update(message);
```

blenk92 commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,148 @@
+    std::vector<uint8_t> encryptedData;
+    try {
+         encCtx->update(message);
+         encryptedData = encCtx->finish();
```
```suggestion
        encryptedData = encCtx->finish();
```

blenk92 commented on examples/key-example.cpp:
```diff
@@ -0,0 +1,49 @@
+
+void createEccKey(openssl::ellipticCurveNid curve, std::string password)
+{
+     auto spec = std::make_shared<ECCSpec>(curve);
```
wrong indentation from here on

blenk92 commented on examples/rsa-example.cpp:
```diff
@@ -0,0 +1,83 @@
+    RSAEncryptionPrivateKeyCtx rsaPrivCtx(*privKey, padding);
+    try {
+        plaintext = rsaPrivCtx.decrypt(ciphertext);
+    }  catch (MoCOCrWException &e) {
```
```suggestion
    } catch (MoCOCrWException &e) {
```
:-D

blenk92 commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,198 @@
+         * DigestType is used as hash function for the padding schemes */
+        ctx = std::make_shared<RSASignaturePrivateKeyCtx>(*privKey, digestType,
+                                                          padding);
+    }  catch (MoCOCrWException &e) {
```
```suggestion
    } catch (MoCOCrWException &e) {
```

blenk92 commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,198 @@
+
+    try {
+        ctx = std::make_shared<RSASignaturePublicKeyCtx>(*pubKey, digestType, rsaPadding);
+    }  catch (MoCOCrWException &e) {
```
```suggestion
    } catch (MoCOCrWException &e) {
```

blenk92 commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,198 @@
+        /* The argument hashFunction is optional. Default is SHA256
+         * The default signature format is ECDSASignatureFormat::ASN1_SEQUENCE_OF_INTS */
+        ctx = std::make_shared<ECDSASignaturePrivateKeyCtx>(*privKey.get(), digestType, sigFormat);
+    }  catch (MoCOCrWException &e) {
```
```suggestion
    } catch (MoCOCrWException &e) {
```

blenk92 commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,198 @@
+    std::shared_ptr<MessageVerificationCtx> ctx(nullptr);
+    try {
+        ctx = std::make_shared<ECDSASignaturePublicKeyCtx>(*pubKey, digestType, sigFormat);
+    }  catch (MoCOCrWException &e) {
```
```suggestion
    } catch (MoCOCrWException &e) {
```

blenk92 commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,198 @@
+    std::shared_ptr<MessageSignatureCtx> ctx(nullptr);
+    try {
+        ctx = std::make_shared<EdDSASignaturePrivateKeyCtx>(*privKey);
+    }  catch (MoCOCrWException &e) {
```
```suggestion
    } catch (MoCOCrWException &e) {
```

blenk92 commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,198 @@
+    std::shared_ptr<MessageVerificationCtx> ctx(nullptr);
+    try {
+        ctx = std::make_shared<EdDSASignaturePublicKeyCtx>(*pubKey);
+    }  catch (MoCOCrWException &e) {
```
```suggestion
    } catch (MoCOCrWException &e) {
```

blenk92 commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,198 @@
+    std::vector<uint8_t> signature;
+    try {
+        signature = signCtx->signMessage(message);
+    }  catch (MoCOCrWException &e) {
```
```suggestion
    } catch (MoCOCrWException &e) {
```

blenk92 commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+    try {
+        authenticatedEncryptor->update(plaintext);
+        ciphertext = authenticatedEncryptor->finish();
+    }  catch (openssl::OpenSSLException &e) {
```
```suggestion
    } catch (openssl::OpenSSLException &e) {
```

blenk92 commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+    try {
+        authenticatedDecryptor->update(aesEncryptResult.ciphertext);
+        authenticatedDecryptor->finish();
+    }  catch (MoCOCrWException &e) {
```
```suggestion
    } catch (MoCOCrWException &e) {
```

blenk92 commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+    try {
+        encryptor->update(plaintext);
+        ciphertext = encryptor->finish();
+    }  catch (openssl::OpenSSLException &e) {
```
```suggestion
    } catch (openssl::OpenSSLException &e) {
```

blenk92 commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+    try {
+        decryptor->update(aesEncryptResult.ciphertext);
+        plaintext = decryptor->finish();
+    }  catch (MoCOCrWException &e) {
```
```suggestion
    } catch (MoCOCrWException &e) {
```

blenk92 commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,213 @@
+    auto ca = getRootCa(signingCert, signingKey);
+    try {
+        cert = std::make_shared<X509Certificate>(ca->signCSR(*csr));
+    }  catch (openssl::OpenSSLException &e) {
```
```suggestion
    } catch (openssl::OpenSSLException &e) {
```

blenk92 commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,213 @@
+    auto ca = getIntermediateCa(signingCert, signingKey);
+    try {
+        cert = std::make_shared<X509Certificate>(ca->signCSR(*csr));
+    }  catch (openssl::OpenSSLException &e) {
```
```suggestion
    } catch (openssl::OpenSSLException &e) {
```

blenk92 commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,148 @@
+    try {
+         encCtx->update(message);
+         encryptedData = encCtx->finish();
+    }  catch (MoCOCrWException &e) {
```
```suggestion
    } catch (MoCOCrWException &e) {
```

blenk92 commented on examples/mac-example.cpp:
```diff
@@ -0,0 +1,67 @@
+
+    try {
+        hmac->verify(authenticationTag);
+    }  catch (MoCOCrWException &e) {
```
```suggestion
    } catch (MoCOCrWException &e) {
```

garw commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+    std::vector<uint8_t> ciphertext;
+};
+
+struct AesAuthenticatedEncryptionResult aesAuthenticatedEncryption()
```
Having the struct keyword here is C, not C++. Please remove.

blenk92 commented on examples/kdf-example.cpp:
```diff
@@ -0,0 +1,62 @@
+{
+    struct KdfData kdfData = {
+        utility::fromHex("deadbeef"),
+        512 / 8,
```
```suggestion
        Hash::getDigestSize(DigestTypes::SHA512),
```
please avoid manual calcualtion of digest length as we already have a function do that (already fixed the same thing in the ecies implementation)

garw commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+            .setPadding(padding)
+            // The IV is optional. Default is a random value
+            .setIV(iv);
+    // the authTagLength is an optional value
```
only needed for authenticated encryption. Defaults to?

garw commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+    encryptorBuilder.setAuthTagLength(authTagLength);
+
+    auto authenticatedEncryptor = encryptorBuilder.buildAuthenticatedEncryptor();
+    authenticatedEncryptor->addAssociatedData(authData);
```
I'd think that this can throw as well? Move into the try {} block?

wusto replied:
It throws depending on the position in the code, but not depending on the input (if you call update or finish beforehand). Therefore I didn't put it into the try {} block. How to proceed here?

Shall I move everything what can throw an exception into a try-catch block?
Up to now I differ between these two cases (position and input dependent)

garw commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+    };
+}
+
+void aesAuthenticatedDecryption(struct AesAuthenticatedEncryptionResult aesEncryptResult)
```
s.a. - the `struct` keyword is not needed in C++ and only adds noise.
Moreover, pass a (const) reference?

garw commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+void aesAuthenticatedDecryption(struct AesAuthenticatedEncryptionResult aesEncryptResult)
+{
+    auto decryptorBuilder = AESCipherBuilder{aeOperationMode, SymmetricCipherKeySize::S_128, aesKey}
+            .setPadding(padding)
```
padding is still optional. IV not.

blenk92 commented on examples/kdf-example.cpp:
```diff
@@ -0,0 +1,62 @@
+void x963kdf(const struct KdfData &kdfData)
+{
+    auto x963kdf = X963KDF(kdfData.digestType);
+    auto derivedKey = x963kdf.deriveKey(kdfData.password, kdfData.outputLength, kdfData.salt);
```
in example i would like to avoid using auto when type is not obvious (such as a return type of a function). If someone that never used that lib needs to understand the code, auto makes it a lot harder.

garw commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+    decryptorBuilder.setAuthTagLength(authTagLength);
+
+    auto authenticatedDecryptor = decryptorBuilder.buildAuthenticatedEncryptor();
+    authenticatedDecryptor->addAssociatedData(authData);
```
s.a. - needs to move into the try {} block?

garw commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+std::vector<uint8_t> aesDecryption(struct AesEncryptionResult aesEncryptResult)
+{
+    auto decryptorBuilder = AESCipherBuilder{plainOperationMode, SymmetricCipherKeySize::S_128, aesKey}
+            .setPadding(padding)
```
add comments here as well?

blenk92 commented on examples/key-example.cpp:
```diff
@@ -0,0 +1,49 @@
+     auto key = AsymmetricKeypair::generate(*spec);
+
+     /* optional: encrypt the key while exporting to PEM */
+     auto pem = key.privateKeyToPem(password);
```
do we also want to show how the get and export the public key?

garw replied:
s.a.

garw commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,167 @@
+        return -1;
+    }
+
+    return 0;
```
Some kind of message also in case of success?

garw commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,213 @@
+#include <mococrw/key_usage.h>
+#include <mococrw/sign_params.h>
+#include <mococrw/csr.h>
+#include <mococrw/openssl_wrap.h>
```
normally there should be no need to include this directly?

blenk92 commented on examples/mac-example.cpp:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    std::cout << "Verification successful." << std::endl;
+
```
usually you don't have an empty line after the last statement of a function

blenk92 commented on examples/mac-example.cpp:
```diff
@@ -0,0 +1,67 @@
+}
+
+std::vector<uint8_t> calculate(const DigestTypes &digestType, const std::vector<uint8_t> &key,
+               const std::vector<uint8_t>& message)
```
wrong indentation

garw commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,213 @@
+static const BasicConstraintsExtension rootCertSignConstraints{true, 0};
+static const auto rootCaKeyUsage = std::make_unique<KeyUsageExtension>(KeyUsageExtension::Builder{}
+                                                                          .keyCertSign().cRLSign().build());
+static const auto rootCaSelfSignParams = std::make_unique<CertificateSigningParameters>(
```
Normally, I would expect that these are generated somewhere close to where you issue the cert - so have them created there instead of a static global. This avoids the need to scroll up and down when reading the example.
Moreover, Why do they need to be a pointer at all? If there is no good reason, don't use pointers.

garw commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,213 @@
+
+
+/******************* Intermediate CA *******************/
+static const auto intermediateCaCertDetails = std::make_unique<DistinguishedName>(
```
same here - move into the creation code and remove pointers?

garw commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,213 @@
+                    *rootCaSelfSignParams));
+}
+
+std::shared_ptr<X509Certificate> rootCaSign(std::shared_ptr<X509Certificate> signingCert,
```
The name is ambigious. Better `signCertificateWithRootCa`

garw replied:
Again: Why does this return a pointer? The library is mostly built so that the objects can and should be used without pointers. Our examples should reflect that.

garw commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,213 @@
+    return cert;
+}
+
+std::shared_ptr<X509Certificate> intermediateCaSign(std::shared_ptr<X509Certificate> signingCert,
```
see comments above for rootCaSign

garw commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,213 @@
+                                                                         .stateOrProvinceName("nebenan")
+                                                                         .serialNumber("08E36DD501941432358AFE8256BC6EFD")
+                                                                         .build());
+static const BasicConstraintsExtension rootCertConstraint{true, 1};
```
These should also move down into the respective CA creation routines and have some comments what is going on here - in particular it should explain that the parameters and how to configure if the issued certificate should be a CA again (ca constraint)

blenk92 commented on examples/rsa-example.cpp:
```diff
@@ -0,0 +1,83 @@
+    auto ciphertext = encrypt_rsa(keyPair, padding, data);
+    auto plaintext = decrypt_rsa(keyPair, padding, ciphertext);
+
+    if (data == plaintext) {
```
do we also want to check ciphertext != plaintext here?

wusto replied:
Do know if this is helpful. This seems like testing if RSA is encrypting the data.

garw commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,148 @@
+using namespace mococrw;
+
+struct EciesEncryptResult {
+    std::vector<uint8_t> ephKey;
```
```suggestion
    std::vector<uint8_t> ephemeralKey;
```
please be verbose with names in the examples. People don't have a VT-100 anymore and read the examples to understand what is going on…

blenk92 commented on src/mac.cpp:
```diff
@@ -26,6 +26,8 @@
 namespace mococrw
 {
 
+using namespace openssl;
```
why do we need this?

garw commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,148 @@
+-----END EC PRIVATE KEY-----
+)";
+
+static const auto digest = DigestTypes::SHA512;
```
As above. Please just define/use these when you instantiate the objects. These lengthy lists of constant require scrolling while trying to understand the example.

wusto replied:
This would require code duplication as these constants are used for encryption and decryption.
Shall I leave it there or duplicate the code for en- and decrypting?

Commit:
Fix src/CMakeLists.txt: remove unused code

The directory cwrap does not exist.

garw commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,148 @@
+    return std::make_unique<mococrw::HMAC>(digest, key);
+});
+
+std::unique_ptr<ECIESEncryptionCtx> buildEncryptionContext()
```
We use this function once. Inline to where used to avoid scrolling? If the boilerplate is removed, this is a 5-liner.

garw commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,148 @@
+            .buildEncryptionCtx(pubKey);
+}
+
+std::unique_ptr<ECIESDecryptionCtx> buildDecryptionContext(const std::vector<uint8_t> &ephKey)
```
In order to make transparent what this requires, please don't use the globals in a function like this but have the private key as an input. This makes it easier to understand the data-flow here. Alternatively, also (like above) inline where called.

blenk92 commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,213 @@
+{
+    return std::make_shared<CertificateAuthority>(
+                *rootCaSignParams,
+                1,
```
maybe we should explain what such magic constans mean here...

Commit:
Make IOUtils.h a private header

IOUtils.h is moved to src/mococrw/private

garw commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,148 @@
+
+int main(void)
+{
+    auto eciesData = encrypt_ecies();
```
Add a comment here that the message serialization of the EC-point, data and MAC is left to the implementer and not specified as part of the usual standards. It depends on the actually implemented protocol.

garw commented on examples/hash-example.cpp:
```diff
@@ -0,0 +1,35 @@
+int main(void)
+{
+    calculateHashSum(DigestTypes::SHA512, utility::fromHex("deadbeef"));
+    return 0;
```
Add a demo of the simple one-off just do it functions. These are probably of more use to most people.

wusto replied:
Done

Commit:
IOUtils: Add new functions

Two functions are added:
- Read private key files in PEM format
- write PEM data to file

garw commented on examples/kdf-example.cpp:
```diff
@@ -0,0 +1,62 @@
+    uint32_t iterations;
+};
+
+void pbkdf2(const struct KdfData &kdfData)
```
can't we just inline this? This would improve readability a lot. This struct just increases the noise.

garw commented on examples/key-example.cpp:
```diff
@@ -0,0 +1,49 @@
+    auto key = AsymmetricKeypair::generate(*spec);
+
+    /* optional: encrypt the key while exporting to PEM */
+    auto pem = key.privateKeyToPem(password);
```
maybe output it? Otherwise exeucting this is quite boring…

garw commented on examples/key-example.cpp:
```diff
@@ -0,0 +1,49 @@
+int main(void)
+{
+    createRsaKey(2048, "secretPassword");
+    createEccKey(openssl::ellipticCurveNid::Ed448, "secretPassword");
```
demonstrate some NIST curves in addition. This also has a slightly different handling

garw commented on examples/rsa-example.cpp:
```diff
@@ -0,0 +1,83 @@
+    RSAEncryptionPrivateKeyCtx rsaPrivCtx(*privKey, padding);
+    try {
+        plaintext = rsaPrivCtx.decrypt(ciphertext);
+    }  catch (MoCOCrWException &e) {
```
This will not throw openssl exceptions?

wusto replied:
The OpenSSLException is caught and forwarded as an MoCOCrWException (asymmetric_crypto_ctx.cpp:168f)

garw commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,198 @@
+     * In MoCOCrW the AsymmetricPrivateKey is a specialisation of AsymmetricPublicKey. Thus
+     * we do an implicit downcast here.
+     */
+    auto rsaVerfiyCtx = getRsaVerifyCtx(rsaPrivKey, padding, digestType);
```
Recommendation from my end would be create function here and put everything in there. This would improve readability.

wusto replied:
I reworked it. Hope that the new solution fits your

garw reviewed (changes requested):
Some general remarks in addition to the comments:

- Please avoid usage of pointers when possible. The library supports that and generally uses this style. Thus, the examples should as well.
- Avoid superfluous use of `struct` keyword in C++
- Think about inlining the smaller functions. This increases readability because it reduces the amount of jumping through the code while reading it.
- Best practice for catching exceptions is `const` ref. Please adapt everywhere.
- Please move the changelog entries into the commits that actually do the changes and not a singular "update changelog" commit.
- Add information about where typos are fixed to the commit message(s).
- Remove internal ticket identifiers from commit messages.

neverpanic commented on CHANGELOG.md:
```diff
@@ -9,6 +9,7 @@ All notable changes to this project will be documented in this file.
 * Replace `explicit_bzero(3)` with `OPENSSL_cleanse(3)` to fix compilation on non-glibc systems. (see #75)
 * Replaced manual calculation of hash digest length by library function in ECIES
 * Improved error message when attempting to create an AES-GCM cipher with empty IV. (see #83)
+* Move DigestTypes from openssl namespace to mococrw namespace. This is an ABI change!
```
Please remove the internal issue reference from the commit message. Also, we usually use present tense imperative, i.e. "Move", not "Moving".

wusto replied:
I don't get want you mean here.
Can you please be more specific?

neverpanic replied:
The commit message contains "Issue: MGUSEC-3493". This should be removed.

Additionally, the commit message should be "DigestTypes: Move from openssl to mococrw namespace".

neverpanic commented on CHANGELOG.md:
```diff
@@ -15,6 +15,8 @@ All notable changes to this project will be documented in this file.
 
 * Exception message in AESCipher (thrown if key has unexpected length)
 * Install target of documentation
+* Add error detection to utility::fromHex (This changes the behaviour of fromHex as it will throw
```
I think this should be in the "changed" section, since it's a change of behavior, not necessarily a bugfix.

neverpanic commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,180 @@
+            // The IV is optional. Default is a random value (Recommendation: use a random value)
+            .setIV(iv);
+    // the authTagLength is an optional value. Default 16 bytes. Only needed of AES AEAD.
+    encryptorBuilder.setAuthTagLength(authTagLength);
```
Any reason why we chain the calls for `.setPadding()` and `.setIV()`, but not the one for `.setAuthTagLength()`?

Usually we don't explicitly name our builder using a variable, but just keep it as a temporary and call `.build()` on it ASAP.

neverpanic commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,180 @@
+            .setPadding(padding)
+            // The IV is optional. Default is a random value (Recommendation: use a random value)
+            .setIV(iv);
+    // the authTagLength is an optional value. Default 16 bytes. Only needed of AES AEAD.
```
```suggestion
    // the authTagLength is an optional value. Default 16 bytes. Only needed by AEAD modes of operation.
```

I think this should be "needed by", not "needed of", and I'd reformulate to "AEAD modes of operation", since technically you could also do this with other algorithms (we just happen to usually use AES).

neverpanic commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,220 @@
+                rootCaSelfSignParams);
+}
+
+X509Certificate singCertificateWithRootCa(const X509Certificate &signingCert,
```
```suggestion
X509Certificate signCertificateWithRootCa(const X509Certificate &signingCert,
```

We should have a certificate singing contest, that could be fun :-D

wusto replied:
:)

neverpanic commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,220 @@
+    }
+}
+
+X509Certificate signCertificateWithIntermediateCa(X509Certificate signingCert,
```
```suggestion
X509Certificate signCertificateWithIntermediateCa(const X509Certificate &signingCert,
```
For consistency, we should use the same types for all of these functions.

neverpanic commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,220 @@
+                                                const AsymmetricKeypair &rootCaKey,
+                                                const X509Certificate &rootCert)
+{
+    auto intermediateCaCertDetails = DistinguishedName(
+                DistinguishedName::Builder{}
+                .organizationalUnitName("Car IT")
+                .organizationName("BMW")
+                .countryName("DE")
+                .commonName("BMW internal CA Certificate").build()
+                );
```
```suggestion
    auto intermediateCaCertDetails =
                DistinguishedName::Builder{}
                .organizationalUnitName("Car IT")
                .organizationName("BMW")
                .countryName("DE")
                .commonName("BMW internal CA Certificate").build();
```

No need to invoke the `DistinguishedName` copy constructor here, from what I can see?

neverpanic commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,220 @@
+                                 const AsymmetricKeypair &signingCaKey,
+                                 const X509Certificate &signingCaCert)
+{
+    auto clientCertDetails = DistinguishedName(
```
Same here.

neverpanic commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,230 @@
+     * Thus it is up to the implementer to define its own protocol (hash-function, key derivation function,
+     * message authentication code, symmetric cipher, key serialization) used for encryption and decryption.
+     * In the current example the following is used:
+     * - SHA512 everywhere where required
```
```suggestion
     * - SHA512 everywhere where a hash function is required
```

neverpanic commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,230 @@
+         * - hybrid
+         */
+        encCtx->getEphemeralKey().toECPoint(openssl::EllipticCurvePointConversionForm::uncompressed),
+                encryptedData,
+                encCtx->getMAC(),
```
I find the indentation a bit unexpected. Any reason the three members of the struct are not indented to the same column?

```suggestion
        encryptedData,
        encCtx->getMAC(),
```

wusto replied:
It's the default of my IDE.
It's fixed now.

neverpanic commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,230 @@
+     * Thus it is up to the user to define its own set of cipher blocks (hash-function, key derivation function,
+     * message authentication code, symmetric cipher) used for encryption and decryption.
+     * In the current example the following is used:
+     * - SHA512 everywhere where required
```
```suggestion
     * - SHA512 everywhere where a hash function is required
```

neverpanic reviewed (changes requested):
Mostly smaller nitpicks now.

garw commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,180 @@
+    std::vector<uint8_t> plaintext;
+    try {
+        decryptor->update(aesEncryptResult.ciphertext);
+        plaintext = decryptor->finish();
```
OpenSSLExceptions are not possible here?

garw commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,220 @@
+
+/******************* Root CA *******************/
+// "Allow" certificate and certificate revocation list signing
+static const auto rootCaKeyUsage = KeyUsageExtension::Builder{}.keyCertSign().cRLSign().build();
```
Any strong reasons why this is still a global while the basic constraint is now a local?

garw commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,220 @@
+CertificateAuthority getRootCa(const X509Certificate &rootCaCert,
+                               const AsymmetricKeypair &rootCaKey)
+{
+    // Set CA to true and path length to 0 (only one intermediate CA)
```
Make clear that we set this as parameters for the signed certificates. I had the impression that this is for the root CA itself and was a bit confused.

garw commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,220 @@
+    // Remark: This will be the X509 content of intermediate CAs not of the root CA itself
+    BasicConstraintsExtension rootCertSignConstraints{true, 0};
+
+    auto rootCaSignParams = CertificateSigningParameters(
```
Outer construction seems superflous. Just calls the copy constructor?

garw commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,220 @@
+    auto intermediateCaCertDetails = DistinguishedName(
+                DistinguishedName::Builder{}
+                .organizationalUnitName("Car IT")
+                .organizationName("BMW")
```
I think the general policy in this lib nowadays is to avoid explicit BMW references

garw commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,220 @@
+}
+
+
+X509Certificate createClientCert(const AsymmetricKeypair &clientKey,
```
I'd split this up. Normally, if you are a CA, you are handed the CSR, do some checks on the csr (e.g. look at the common name) and then sign it. You're usually NOT handed the client key.

garw commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,230 @@
+    const int digestLength = Hash::getDigestSize(digest);
+    const auto kdfFunc = std::make_shared<X963KDF>(digest);
+    /* Dependency injection function, as the key for MAC is generated based on a random ECC key in the ECIES class */
+    const auto macFunc = std::function<std::unique_ptr<MessageAuthenticationCode>(const std::vector<uint8_t>&)>(
```
conversion to std::function shouldn't be necessary here.

garw commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,230 @@
+
+    /* Get the encryption context */
+    ECIESCtxBuilder encBuilder;
+    std::unique_ptr<ECIESEncryptionCtx> encCtx = encBuilder
```
I'd really just put all of the definitions above into the builder here. This allows the reader to just read all of the definitions top to bottom without jumping above.

garw commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,230 @@
+    try {
+        encCtx->update(message);
+        encryptedData = encCtx->finish();
+    } catch (const MoCOCrWException &e) {
```
no openssl exceptions? if yes, add a comment.

garw commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,230 @@
+    };
+}
+
+void decrypt_ecies(EciesEncryptResult eciesData)
```
```suggestion
void decrypt_ecies(const EciesEncryptResult &eciesData)
```
?

garw commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,230 @@
+
+    /* Get the decryption context */
+    ECIESCtxBuilder decBuilder;
+    auto decCtx = decBuilder
```
s.a. regarding inlining the config.
Moreover, there seems to be no need to actually have a named builder object.

garw commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,230 @@
+
+        /* Get the plaintext and verify the MAC */
+        decryptedData = decCtx->finish();
+    } catch (const MoCOCrWException &e) {
```
s.a. regarding OpenSSLException

garw commented on examples/kdf-example.cpp:
```diff
@@ -0,0 +1,44 @@
+{
+
+    auto password = utility::fromHex("deadbeef");
+    auto outputLength= Hash::getDigestSize(DigestTypes::SHA512);
```
Why is this set to the length of SHA512? I think this is a bit confusing. I typical use-case is probably rather that the output is 128 or 256bit for an AES key? Setting it to the size of a sha hash create a confusion relation imho.

garw commented on examples/key-example.cpp:
```diff
@@ -0,0 +1,69 @@
+    return key;
+}
+
+AsymmetricKeypair createEccKey(openssl::ellipticCurveNid curve, std::string password)
```
```suggestion
AsymmetricKeypair createEccKey(openssl::ellipticCurveNid curve, const std::string &password)
```

garw commented on examples/key-example.cpp:
```diff
@@ -0,0 +1,69 @@
+
+using namespace mococrw;
+
+AsymmetricKeypair createRsaKey(uint keySize, std::string password)
```
```suggestion
AsymmetricKeypair createRsaKey(uint keySize, const std::string &password)
```

garw commented on examples/key-example.cpp:
```diff
@@ -0,0 +1,69 @@
+    exportPublicKey(rsaKeyPair);
+
+    auto eccKeyPair = createEccKey(openssl::ellipticCurveNid::Ed448, "secretPassword");
+    /* see comment above */
```
```suggestion
```
Well, I still remember 2 lines before…

garw commented on examples/key-example.cpp:
```diff
@@ -0,0 +1,69 @@
+    exportPublicKey(eccKeyPair);
+
+    auto eccNistKeyPair = createEccKey(openssl::ellipticCurveNid::SECP_521r1, "secretPassword");
+    /* see comment above */
```
```suggestion
```

garw commented on examples/mac-example.cpp:
```diff
@@ -0,0 +1,67 @@
+
+using namespace mococrw;
+
+std::unique_ptr<mococrw::HMAC> doHmac(DigestTypes digestType, const std::vector<uint8_t> &key,
```
just inline this to the places where its used? 
Would defininately eliminate the need for a pointer and is just 2 lines instead of 1. But will most probably be used like this in almost all of the cases.

garw commented on examples/mac-example.cpp:
```diff
@@ -0,0 +1,67 @@
+    std::vector<uint8_t> message = utility::fromHex("deadbeef");
+    std::vector<uint8_t> key = utility::fromHex("beefdead");
+
+    auto tag = calculate(digestType, key, message);
```
Just demo this here all at once. It is in principle 3 lines. Now I need to read 2 addional functions to understand 3 LOC

garw commented on examples/rsa-example.cpp:
```diff
@@ -0,0 +1,90 @@
+    auto keyPair = std::make_shared<AsymmetricKeypair>(AsymmetricKeypair::generateRSA());
+    std::vector<uint8_t> data = utility::fromHex("deadbeef");
+    DigestTypes oaepDigestType = DigestTypes::SHA512;
+    // you can use a different hash function for MGF1 than for OAEP padding if you want/need to
```
```suggestion
    // you can use a different hash function for MGF1 than for OAEP padding if you want/need to
    // beware that this is really uncommon.
```

garw commented on examples/rsa-example.cpp:
```diff
@@ -0,0 +1,90 @@
+    DigestTypes mgfDigestType = DigestTypes::SHA3_256;
+    std::string oaepLabel = "this is my awesome label";
+
+    auto maskGenerationFunction = std::make_shared<MGF1>(MGF1(mgfDigestType));
```
just add constant and comment inline here?

garw commented on examples/rsa-example.cpp:
```diff
@@ -0,0 +1,90 @@
+
+    auto maskGenerationFunction = std::make_shared<MGF1>(MGF1(mgfDigestType));
+
+    /* no argument to OAEPPadding is mandatory:
+     * - default hashFunction = SHA256
+     * - default maskGenerationFunction = MGF1 with SHA256 (hashFunction)
+     * - default label = "" (empty string)
+     */
```
```suggestion
    /* no argument to OAEPPadding is mandatory:
     * - default hashFunction = SHA256
     * (normal use-cases do not specify more than the hash function)
     * - default maskGenerationFunction = MGF1 with the hash function specified in parameter hashFunction
     * - default label = "" (empty string)
     */
```

garw commented on examples/rsa-example.cpp:
```diff
@@ -0,0 +1,90 @@
+    return ciphertext;
+}
+
+std::vector<uint8_t> decrypt_rsa(const std::shared_ptr<AsymmetricPrivateKey> &privKey,
```
consistency with signature example: Below we handle keys as const ref to the key object.

garw commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,229 @@
+
+std::vector<uint8_t> rsaSign(const AsymmetricPrivateKey &privKey,
+                             const DigestTypes digestType,
+                             const std::shared_ptr<RSASignaturePadding> padding,
```
const ref vs non-const noref?

garw commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,229 @@
+                             const std::shared_ptr<RSASignaturePadding> padding,
+                             const std::vector<uint8_t> &message)
+{
+    std::shared_ptr<MessageSignatureCtx> signCtx(nullptr);
```
nullptr is the default constructor.

garw commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,229 @@
+                 const std::vector<uint8_t> &signature,
+                 const std::vector<uint8_t> &message)
+{
+    std::shared_ptr<MessageVerificationCtx> verifyCtx(nullptr);
```
```suggestion
    std::shared_ptr<MessageVerificationCtx> verifyCtx;
```

garw commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,229 @@
+std::vector<uint8_t> edDsaSign(const AsymmetricPrivateKey &privKey,
+                               const std::vector<uint8_t> &message)
+{
+    std::shared_ptr<MessageSignatureCtx> signCtx(nullptr);
```
```suggestion
    std::shared_ptr<MessageSignatureCtx> signCtx;
```

garw commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,229 @@
+                 const std::vector<uint8_t> &signature,
+                 const std::vector<uint8_t> &message)
+{
+    std::shared_ptr<MessageVerificationCtx> verifyCtx(nullptr);
```
```suggestion
    std::shared_ptr<MessageVerificationCtx> verifyCtx;
```

garw commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,229 @@
+    auto mgf1DigestType = DigestTypes::SHA256;
+    auto rsaSignatureDigestType = DigestTypes::SHA512;
+    auto mgf1 = std::make_shared<MGF1>(MGF1(mgf1DigestType));
+    auto padding = std::make_shared<PSSPadding>(PSSPadding(mgf1, mococrw::Hash::getDigestSize(rsaSignatureDigestType)));
```
The only parameter that is unclear from reading the input because its just a number (the salt length) is unnamed. If you want to inline things, inline the MGF but not the salt length because I need to take a look at the function definition to understand what the second parameter is… Better would be to have a variable for the things for which its not clear from the type what their meaning is.

garw commented on tests/unit/test_util.cpp:
```diff
@@ -53,3 +53,30 @@ TEST_F(UtilTest, toHexWithConsequentZeroes)
+    ASSERT_TRUE(result.empty());
+
+    ASSERT_NO_THROW(result = mococrw::utility::fromHex("0x1"));
+    ASSERT_TRUE(result == std::vector<uint8_t>{1});
```
```suggestion
    ASSERT_THAT(result, Eq(std::vector<uint8_t>{1}));
```
Please use the feature test macros of gtest that give you reasonable output in case of mismatches. Same for all equality tests below.

garw reviewed (dismissed):
Some (minor) quality issues left to be resolved.
In general, the examples still have a bit of an inconsistent handling w.r.t. OpenSSLExceptions. Some of the examples always catch both. Some only MoCoCrWExceptions. This needs a bit of comments for a reader to understand when they're expect to catch both and under which circumstances OpenSSLExceptions shouldn't happen. Maybe it is sufficient in these cases to just globally catch them and indicate that that something is wrong?

neverpanic commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,211 @@
+            .setPadding(padding)
+            // The IV is optional. Default is a random value (Recommendation: use a random value)
+            .setIV(iv)
+            // the authTagLength is an optional value. Default 16 bytes. Only needed of AES AEAD.
```
```suggestion
            // the authTagLength is an optional value. Default 16 bytes. Only needed by AES AEAD.
```

I think this should be "needed by" or "needed for".

neverpanic commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,211 @@
+            // The IV is optional. Default is a random value (Recommendation: use a random value)
+            .setIV(iv);
+
+    auto encryptor = encryptorBuilder.buildEncryptor();
```
I would prefer of this would be chained to the encryptorBuilder calls, like in the other examples.

neverpanic commented on examples/aes-example.cpp:
```diff
@@ -0,0 +1,211 @@
+            // mandatory value for decryption. Don't forget to set it. Set the same value as set during encryption
+            .setIV(aesEncryptResult.iv);
+
+    auto decryptor = decryptorBuilder.buildDecryptor();
```
Same applies here.

neverpanic commented on examples/ca-csr-example.cpp:
```diff
@@ -0,0 +1,234 @@
+            .addExtension(rootCaKeyUsage)
+            .build();
+
+    // These are all supported X509 cert details supported by mococrw
```
"supported" occurs twice in this comment.

neverpanic commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,265 @@
+
+    /* Get the encryption context */
+    ECIESCtxBuilder encBuilder;
+    std::unique_ptr<ECIESEncryptionCtx> encCtx = encBuilder
```
```suggestion
    std::unique_ptr<ECIESEncryptionCtx> encCtx = ECIESCtxBuilder{}
```

Could we create an anonymous temporary ECIESCtxBuilder object here, rather than explicitly assigning it to a variable?

neverpanic commented on examples/ecies-example.cpp:
```diff
@@ -0,0 +1,265 @@
+        exit(EXIT_FAILURE);
+    }
+
+    /* The underlying openssl function recoginzes the used serialization format of the ephemeral key reading the first
```
```suggestion
    /* The underlying openssl function recognizes the used serialization format of the ephemeral key reading the first
```
Typo.

neverpanic commented on examples/rsa-example.cpp:
```diff
@@ -0,0 +1,95 @@
+
+int main(void)
+{
+    auto keyPair = std::make_shared<AsymmetricKeypair>(AsymmetricKeypair::generateRSA());
```
An `AsymmetricKeypair` is a wrapper around a `std::shared_ptr` and thus lightweight. We should copy it directly rather than creating shared pointers to it. This also applies to the `encrypt_rsa` and `decrypt_rsa` interfaces above.

neverpanic commented on examples/sig-example.cpp:
```diff
@@ -0,0 +1,255 @@
+    } catch (const MoCOCrWException &e) {
+        /* Possible reasons:
+         * - error in openssl (sign, padding, ...)
+         * - Hash function's digest size doesn't match the message's digest size
```
Should this list (and the equivalents in the other verify functions) include "incorrect signature"?

wusto replied:
Yes. You are right. Thanks!

neverpanic reviewed (changes requested):
A couple of minor points. Almost there!

neverpanic reviewed (approved):
LG2M
